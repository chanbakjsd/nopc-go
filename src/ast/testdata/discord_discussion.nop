// unclear
#no_std
#no_auto_free

// unclear
import strconv

struct Point {
    pub x, y i32 // Types don't matter in antlr files for now
}

fn main() {
    // change to dot if no double colon
    let pointA = Point.new(123, 456)

    let mut pointB = pointA.clone()
    pointB.translate(5, 3)
    if pointB.x < 100 {
        println("Should never execute")
    } else {
        println("Should always execute")
    }
    println(pointB) //(128, 459)
}

impl Point {
    fn new(x, y i32) Point {
        return Point{x, y}
    }
    fn translate(*mut self, x, y i32) {
        self->x += x
        self->y += y
    }
}

impl Stringable for Point {
    fn string(*mut self) string {
         return "(" + strconv.Atoi(self->x) + ", " + strconv.Atoi(self->y) + ")"
    }
}

//In internal compiler files
trait Stringable {
    fn string() string
}
