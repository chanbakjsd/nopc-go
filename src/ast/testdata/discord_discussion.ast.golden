{
	"Source": "testdata/discord_discussion.nop",
	"CompilerPragma": [
		"#no_std",
		"#no_auto_free"
	],
	"Import": [
		"strconv"
	],
	"Structs": [
		{
			"Public": false,
			"Name": "Point",
			"FieldNames": [
				"x",
				"y"
			],
			"FieldTypes": [
				{
					"Name": "i32",
					"ToInferType": false,
					"InnerType": null,
					"Public": true,
					"Static": false,
					"Mutable": false
				},
				{
					"Name": "i32",
					"ToInferType": false,
					"InnerType": null,
					"Public": true,
					"Static": false,
					"Mutable": false
				}
			]
		}
	],
	"Traits": [
		{
			"Name": "Stringable",
			"Functions": [
				{
					"Name": "string",
					"Parameters": [],
					"ReturnType": {
						"Name": "string",
						"ToInferType": false,
						"InnerType": null,
						"Public": false,
						"Static": false,
						"Mutable": false
					},
					"Body": null
				}
			]
		}
	],
	"Impls": [
		{
			"TraitName": "Point",
			"TypeName": {
				"Name": "",
				"ToInferType": false,
				"InnerType": null,
				"Public": false,
				"Static": false,
				"Mutable": false
			},
			"Implementations": [
				{
					"Name": "new",
					"Parameters": [
						{
							"Name": "x",
							"Type": {
								"Name": "i32",
								"ToInferType": false,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							}
						},
						{
							"Name": "y",
							"Type": {
								"Name": "i32",
								"ToInferType": false,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							}
						}
					],
					"ReturnType": {
						"Name": "Point",
						"ToInferType": false,
						"InnerType": null,
						"Public": false,
						"Static": false,
						"Mutable": false
					},
					"Body": [
						{
							"Type": "[ReturnStatement]",
							"Actual": {
								"Type": "[ObjectInitializer]",
								"ResolvedType": {
									"Name": "",
									"ToInferType": false,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": null,
								"Arguments": [
									{
										"Type": "[Type]",
										"ResolvedType": {
											"Name": "Point",
											"ToInferType": false,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": null,
										"Arguments": null
									},
									{
										"Type": "[PropertyName]",
										"ResolvedType": {
											"Name": "",
											"ToInferType": true,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": "x",
										"Arguments": null
									},
									{
										"Type": "[PropertyName]",
										"ResolvedType": {
											"Name": "",
											"ToInferType": true,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": "y",
										"Arguments": null
									}
								]
							}
						}
					]
				},
				{
					"Name": "translate",
					"Parameters": [
						{
							"Name": "self",
							"Type": {
								"Name": "self",
								"ToInferType": true,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": true
							}
						},
						{
							"Name": "x",
							"Type": {
								"Name": "i32",
								"ToInferType": false,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							}
						},
						{
							"Name": "y",
							"Type": {
								"Name": "i32",
								"ToInferType": false,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							}
						}
					],
					"ReturnType": {
						"Name": "",
						"ToInferType": false,
						"InnerType": null,
						"Public": false,
						"Static": false,
						"Mutable": false
					},
					"Body": [
						{
							"Type": "[Assignment]",
							"Actual": {
								"Target": {
									"Type": "[PointerAccess]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[Self]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": false,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": null,
											"Arguments": null
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "x",
											"Arguments": null
										}
									]
								},
								"Source": {
									"Type": "[Addition]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PointerAccess]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": null,
											"Arguments": [
												{
													"Type": "[Self]",
													"ResolvedType": {
														"Name": "",
														"ToInferType": false,
														"InnerType": null,
														"Public": false,
														"Static": false,
														"Mutable": false
													},
													"LiteralValue": null,
													"Arguments": null
												},
												{
													"Type": "[PropertyName]",
													"ResolvedType": {
														"Name": "",
														"ToInferType": true,
														"InnerType": null,
														"Public": false,
														"Static": false,
														"Mutable": false
													},
													"LiteralValue": "x",
													"Arguments": null
												}
											]
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "x",
											"Arguments": null
										}
									]
								}
							}
						},
						{
							"Type": "[Assignment]",
							"Actual": {
								"Target": {
									"Type": "[PointerAccess]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[Self]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": false,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": null,
											"Arguments": null
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "y",
											"Arguments": null
										}
									]
								},
								"Source": {
									"Type": "[Addition]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PointerAccess]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": null,
											"Arguments": [
												{
													"Type": "[Self]",
													"ResolvedType": {
														"Name": "",
														"ToInferType": false,
														"InnerType": null,
														"Public": false,
														"Static": false,
														"Mutable": false
													},
													"LiteralValue": null,
													"Arguments": null
												},
												{
													"Type": "[PropertyName]",
													"ResolvedType": {
														"Name": "",
														"ToInferType": true,
														"InnerType": null,
														"Public": false,
														"Static": false,
														"Mutable": false
													},
													"LiteralValue": "y",
													"Arguments": null
												}
											]
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "y",
											"Arguments": null
										}
									]
								}
							}
						}
					]
				}
			]
		},
		{
			"TraitName": "Stringable",
			"TypeName": {
				"Name": "Point",
				"ToInferType": false,
				"InnerType": null,
				"Public": false,
				"Static": false,
				"Mutable": false
			},
			"Implementations": [
				{
					"Name": "string",
					"Parameters": [
						{
							"Name": "self",
							"Type": {
								"Name": "self",
								"ToInferType": true,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": true
							}
						}
					],
					"ReturnType": {
						"Name": "string",
						"ToInferType": false,
						"InnerType": null,
						"Public": false,
						"Static": false,
						"Mutable": false
					},
					"Body": [
						{
							"Type": "[ReturnStatement]",
							"Actual": {
								"Type": "[Addition]",
								"ResolvedType": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": null,
								"Arguments": [
									{
										"Type": "[Addition]",
										"ResolvedType": {
											"Name": "",
											"ToInferType": true,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": null,
										"Arguments": [
											{
												"Type": "[Addition]",
												"ResolvedType": {
													"Name": "",
													"ToInferType": true,
													"InnerType": null,
													"Public": false,
													"Static": false,
													"Mutable": false
												},
												"LiteralValue": null,
												"Arguments": [
													{
														"Type": "[Addition]",
														"ResolvedType": {
															"Name": "",
															"ToInferType": true,
															"InnerType": null,
															"Public": false,
															"Static": false,
															"Mutable": false
														},
														"LiteralValue": null,
														"Arguments": [
															{
																"Type": "[Literal]",
																"ResolvedType": {
																	"Name": "ustring",
																	"ToInferType": false,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": "(",
																"Arguments": null
															},
															{
																"Type": "[FunctionCall]",
																"ResolvedType": {
																	"Name": "",
																	"ToInferType": false,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": null,
																"Arguments": [
																	{
																		"Type": "[Property]",
																		"ResolvedType": {
																			"Name": "",
																			"ToInferType": true,
																			"InnerType": null,
																			"Public": false,
																			"Static": false,
																			"Mutable": false
																		},
																		"LiteralValue": null,
																		"Arguments": [
																			{
																				"Type": "[PropertyName]",
																				"ResolvedType": {
																					"Name": "",
																					"ToInferType": true,
																					"InnerType": null,
																					"Public": false,
																					"Static": false,
																					"Mutable": false
																				},
																				"LiteralValue": "strconv",
																				"Arguments": null
																			},
																			{
																				"Type": "[PropertyName]",
																				"ResolvedType": {
																					"Name": "",
																					"ToInferType": true,
																					"InnerType": null,
																					"Public": false,
																					"Static": false,
																					"Mutable": false
																				},
																				"LiteralValue": "Atoi",
																				"Arguments": null
																			}
																		]
																	},
																	{
																		"Type": "[PointerAccess]",
																		"ResolvedType": {
																			"Name": "",
																			"ToInferType": true,
																			"InnerType": null,
																			"Public": false,
																			"Static": false,
																			"Mutable": false
																		},
																		"LiteralValue": null,
																		"Arguments": [
																			{
																				"Type": "[Self]",
																				"ResolvedType": {
																					"Name": "",
																					"ToInferType": false,
																					"InnerType": null,
																					"Public": false,
																					"Static": false,
																					"Mutable": false
																				},
																				"LiteralValue": null,
																				"Arguments": null
																			},
																			{
																				"Type": "[PropertyName]",
																				"ResolvedType": {
																					"Name": "",
																					"ToInferType": true,
																					"InnerType": null,
																					"Public": false,
																					"Static": false,
																					"Mutable": false
																				},
																				"LiteralValue": "x",
																				"Arguments": null
																			}
																		]
																	}
																]
															}
														]
													},
													{
														"Type": "[Literal]",
														"ResolvedType": {
															"Name": "ustring",
															"ToInferType": false,
															"InnerType": null,
															"Public": false,
															"Static": false,
															"Mutable": false
														},
														"LiteralValue": ", ",
														"Arguments": null
													}
												]
											},
											{
												"Type": "[FunctionCall]",
												"ResolvedType": {
													"Name": "",
													"ToInferType": false,
													"InnerType": null,
													"Public": false,
													"Static": false,
													"Mutable": false
												},
												"LiteralValue": null,
												"Arguments": [
													{
														"Type": "[Property]",
														"ResolvedType": {
															"Name": "",
															"ToInferType": true,
															"InnerType": null,
															"Public": false,
															"Static": false,
															"Mutable": false
														},
														"LiteralValue": null,
														"Arguments": [
															{
																"Type": "[PropertyName]",
																"ResolvedType": {
																	"Name": "",
																	"ToInferType": true,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": "strconv",
																"Arguments": null
															},
															{
																"Type": "[PropertyName]",
																"ResolvedType": {
																	"Name": "",
																	"ToInferType": true,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": "Atoi",
																"Arguments": null
															}
														]
													},
													{
														"Type": "[PointerAccess]",
														"ResolvedType": {
															"Name": "",
															"ToInferType": true,
															"InnerType": null,
															"Public": false,
															"Static": false,
															"Mutable": false
														},
														"LiteralValue": null,
														"Arguments": [
															{
																"Type": "[Self]",
																"ResolvedType": {
																	"Name": "",
																	"ToInferType": false,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": null,
																"Arguments": null
															},
															{
																"Type": "[PropertyName]",
																"ResolvedType": {
																	"Name": "",
																	"ToInferType": true,
																	"InnerType": null,
																	"Public": false,
																	"Static": false,
																	"Mutable": false
																},
																"LiteralValue": "y",
																"Arguments": null
															}
														]
													}
												]
											}
										]
									},
									{
										"Type": "[Literal]",
										"ResolvedType": {
											"Name": "ustring",
											"ToInferType": false,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": ")",
										"Arguments": null
									}
								]
							}
						}
					]
				}
			]
		}
	],
	"Globals": null,
	"Constants": null,
	"Functions": [
		{
			"Name": "main",
			"Parameters": [],
			"ReturnType": {
				"Name": "",
				"ToInferType": false,
				"InnerType": null,
				"Public": false,
				"Static": false,
				"Mutable": false
			},
			"Body": [
				{
					"Type": "[VariableDecl]",
					"Actual": {
						"Targets": [
							{
								"Name": "pointA",
								"Type": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								}
							}
						],
						"DefaultValue": {
							"Type": "[FunctionCall]",
							"ResolvedType": {
								"Name": "",
								"ToInferType": false,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							},
							"LiteralValue": null,
							"Arguments": [
								{
									"Type": "[Property]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "Point",
											"Arguments": null
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "new",
											"Arguments": null
										}
									]
								},
								{
									"Type": "[Literal]",
									"ResolvedType": {
										"Name": "i64",
										"ToInferType": false,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": 123,
									"Arguments": null
								},
								{
									"Type": "[Literal]",
									"ResolvedType": {
										"Name": "i64",
										"ToInferType": false,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": 456,
									"Arguments": null
								}
							]
						},
						"IsConstant": false
					}
				},
				{
					"Type": "[VariableDecl]",
					"Actual": {
						"Targets": [
							{
								"Name": "pointB",
								"Type": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": true
								}
							}
						],
						"DefaultValue": {
							"Type": "[Property]",
							"ResolvedType": {
								"Name": "",
								"ToInferType": true,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							},
							"LiteralValue": null,
							"Arguments": [
								{
									"Type": "[PropertyName]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": "pointA",
									"Arguments": null
								},
								{
									"Type": "[PropertyName]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": "clone",
									"Arguments": null
								}
							]
						},
						"IsConstant": false
					}
				},
				{
					"Type": "[EvaluatedStatement]",
					"Actual": {
						"Type": "[FunctionCall]",
						"ResolvedType": {
							"Name": "",
							"ToInferType": false,
							"InnerType": null,
							"Public": false,
							"Static": false,
							"Mutable": false
						},
						"LiteralValue": null,
						"Arguments": [
							{
								"Type": "[Property]",
								"ResolvedType": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": null,
								"Arguments": [
									{
										"Type": "[PropertyName]",
										"ResolvedType": {
											"Name": "",
											"ToInferType": true,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": "pointB",
										"Arguments": null
									},
									{
										"Type": "[PropertyName]",
										"ResolvedType": {
											"Name": "",
											"ToInferType": true,
											"InnerType": null,
											"Public": false,
											"Static": false,
											"Mutable": false
										},
										"LiteralValue": "translate",
										"Arguments": null
									}
								]
							},
							{
								"Type": "[Literal]",
								"ResolvedType": {
									"Name": "i64",
									"ToInferType": false,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": 5,
								"Arguments": null
							},
							{
								"Type": "[Literal]",
								"ResolvedType": {
									"Name": "i64",
									"ToInferType": false,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": 3,
								"Arguments": null
							}
						]
					}
				},
				{
					"Type": "[IfStatement]",
					"Actual": {
						"Condition": {
							"Type": "\u003c",
							"ResolvedType": {
								"Name": "",
								"ToInferType": true,
								"InnerType": null,
								"Public": false,
								"Static": false,
								"Mutable": false
							},
							"LiteralValue": null,
							"Arguments": [
								{
									"Type": "[Property]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": true,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "pointB",
											"Arguments": null
										},
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "x",
											"Arguments": null
										}
									]
								},
								{
									"Type": "[Literal]",
									"ResolvedType": {
										"Name": "i64",
										"ToInferType": false,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": 100,
									"Arguments": null
								}
							]
						},
						"IfBlock": [
							{
								"Type": "[EvaluatedStatement]",
								"Actual": {
									"Type": "[FunctionCall]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": false,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "println",
											"Arguments": null
										},
										{
											"Type": "[Literal]",
											"ResolvedType": {
												"Name": "ustring",
												"ToInferType": false,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "Should never execute",
											"Arguments": null
										}
									]
								}
							}
						],
						"ElseBlock": [
							{
								"Type": "[EvaluatedStatement]",
								"Actual": {
									"Type": "[FunctionCall]",
									"ResolvedType": {
										"Name": "",
										"ToInferType": false,
										"InnerType": null,
										"Public": false,
										"Static": false,
										"Mutable": false
									},
									"LiteralValue": null,
									"Arguments": [
										{
											"Type": "[PropertyName]",
											"ResolvedType": {
												"Name": "",
												"ToInferType": true,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "println",
											"Arguments": null
										},
										{
											"Type": "[Literal]",
											"ResolvedType": {
												"Name": "ustring",
												"ToInferType": false,
												"InnerType": null,
												"Public": false,
												"Static": false,
												"Mutable": false
											},
											"LiteralValue": "Should always execute",
											"Arguments": null
										}
									]
								}
							}
						]
					}
				},
				{
					"Type": "[EvaluatedStatement]",
					"Actual": {
						"Type": "[FunctionCall]",
						"ResolvedType": {
							"Name": "",
							"ToInferType": false,
							"InnerType": null,
							"Public": false,
							"Static": false,
							"Mutable": false
						},
						"LiteralValue": null,
						"Arguments": [
							{
								"Type": "[PropertyName]",
								"ResolvedType": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": "println",
								"Arguments": null
							},
							{
								"Type": "[PropertyName]",
								"ResolvedType": {
									"Name": "",
									"ToInferType": true,
									"InnerType": null,
									"Public": false,
									"Static": false,
									"Mutable": false
								},
								"LiteralValue": "pointB",
								"Arguments": null
							}
						]
					}
				}
			]
		}
	]
}